cpp/src/parquet/encoding.cc

```diff
@@ -3037,11 +3073,180 @@ class RleBooleanDecoder : public DecoderImpl, virtual public BooleanDecoder {
+        },
+        []() { return Status::OK(); }));
+    last_value_ = last_value_view;
+  }
+
+  ::arrow::BufferBuilder sink_;
+  DeltaBitPackEncoder<Int32Type> prefix_length_encoder_;
+  DeltaLengthByteArrayEncoder<ByteArrayType> suffix_encoder_;
+  std::string last_value_;
+};
+
+template <typename DType>
+void DeltaByteArrayEncoder<DType>::Put(const T* src, int num_values) {
+  if (num_values == 0) {
+    return;
+  }
+  ArrowPoolVector<int32_t> prefix_lengths(num_values,
+                                          ::arrow::stl::allocator<int32_t>(pool_));
+  std::string_view last_value_view = last_value_;
+
+  int i = 0;
+  while (i < num_values) {
+    // Convert to ByteArray so we can pass to the suffix_encoder_.
+    auto value = reinterpret_cast<const ByteArray*>(&src[i]);
+    if (ARROW_PREDICT_FALSE(value->len >= kMaxByteArraySize)) {
+      throw Status::Invalid("Parquet cannot store strings with size 2GB or more");
+    }
+
+    auto view = string_view{reinterpret_cast<const char*>(value->ptr), value->len};
+    uint32_t j = 0;
+    while (j < std::min(value->len, static_cast<uint32_t>(last_value_view.length()))) {
+      if (last_value_view[j] != view[j]) {
+        break;
+      }
+      j++;
+    }
+
+    prefix_lengths[i] = j;
+    const uint8_t* suffix_ptr = value->ptr + j;
+    const uint32_t suffix_length = static_cast<uint32_t>(value->len - j);
+    last_value_view =
+        string_view{reinterpret_cast<const char*>(suffix_ptr), suffix_length};
+    // Convert suffix to ByteArray so it can be passed to the suffix_encoder_.
+    const ByteArray suffix(suffix_length, suffix_ptr);
+    suffix_encoder_.Put(&suffix, 1);
+    i++;
+  }
+  prefix_length_encoder_.Put(prefix_lengths.data(), num_values);
+  last_value_ = last_value_view;
+}
+
+template <typename DType>
+void DeltaByteArrayEncoder<DType>::Put(const ::arrow::Array& values) {
+  if (::arrow::is_binary_like(values.type_id())) {
+    PutBinaryArray(checked_cast<const ::arrow::BinaryArray&>(values));
+  } else if (::arrow::is_large_binary_like(values.type_id())) {
+    PutBinaryArray(checked_cast<const ::arrow::LargeBinaryArray&>(values));
+  } else if (::arrow::is_fixed_size_binary(values.type_id())) {
+    PutBinaryArray(checked_cast<const ::arrow::FixedSizeBinaryArray&>(values));
+  } else {
+    throw ParquetException("Only BaseBinaryArray and subclasses supported");
+  }
+}
+
+template <typename DType>
+std::shared_ptr<Buffer> DeltaByteArrayEncoder<DType>::FlushValues() {
+  PARQUET_THROW_NOT_OK(sink_.Resize(EstimatedDataEncodedSize(), false));
+
+  std::shared_ptr<Buffer> prefix_lengths = prefix_length_encoder_.FlushValues();
+  PARQUET_THROW_NOT_OK(sink_.Append(prefix_lengths->data(), prefix_lengths->size()));
+
+  std::shared_ptr<Buffer> suffixes = suffix_encoder_.FlushValues();
+  PARQUET_THROW_NOT_OK(sink_.Append(suffixes->data(), suffixes->size()));
+
+  std::shared_ptr<Buffer> buffer;
+  PARQUET_THROW_NOT_OK(sink_.Finish(&buffer, true));
+  return buffer;
```

Comment:
Should clear context here.

```suggestion
  last_value_.clear();
  return buffer;
```